### DISCUSSION TITLE
protoc-c creates a name conflict in a proto3 file

### DISCUSSION THREAD

**Discussion by oleg-kiselev** [protobuf-c/protobuf-c#637] [Q&A]
I am trying to compile a 3rd party .proto files into C.  I am running into an issue where a message contains the following fields:
```
    bool has_minimum = 5;
    int64 minimum = 6;
```

protoc generates:
```
  protobuf_c_boolean has_has_minimum;
  protobuf_c_boolean has_minimum;
  protobuf_c_boolean has_minimum;
  int64_t minimum;
```

I can not change the 3rd party .proto file to avoid this collision.  The simplest solution is probably changing protoc-c to generate a boolean field name that is less likely to conflict with the fields in .proto files.  Something elaborate and unlikely, like a `__protoc_c_exists__` suffix instead of `has_` prefix?

**Comment by edmonds** [comment]
> The simplest solution is probably changing protoc-c to generate a boolean field name that is less likely to conflict with the fields in .proto files.

This would be an API break for every program that ever used optional fields with `protobuf-c`, because their `has_*` fields would be gone.

If someone actually uses `has_` as a field name prefix in a protobuf message definition it would have to be escaped somehow. Maybe we escape `has_` to `has_PROTOBUF_C_VERY_INCONVENIENT_NAME_` and hope no one uses *that* as a prefix.

**Comment by oleg-kiselev** [comment]
For my very narrow needs, the issue stems from all fields in this message being optional, as seems to be the norm for protobuf 3.  

In this .proto the `bool has_minimum` field is there to differentiate between an `int64 minimum` being a default "0" value because it wasn't set at all vs. being 0 because it was set to 0.  It isn't clear to me why they aren't defaulting `minimum` to LLONG_MIN instead of using a separate boolean flag.  I suppose they had their reasons?   This field naming works fine with the main protobufs 3 protoc for C++ and python, so the 3rd party in this case had no reason to expect an issue.  Nor will they have any motivation to change their code to accommodate a C user, since their exclusive focus is on C++ and python.

> This would be an API break for every program that ever used optional fields with protobuf-c, because their has_* fields would be gone.

Yes, starting with the unit tests in protoc itself...

It's unclear to me if you are proposing modifying the message's "primary" field, or the optional field's "has_" boolean flag.   Renaming either field is going to trip the user code that is parsing the message.   It's easy to constrain the changes to only the fields that have the `has_` prefix.  This will avoid breaking the protoc-c's own unit tests.   Augmenting the protoc's generated field names for _this specific case_ with the "unlikely" string will work in _this case_: the end-user code expects to have all of the "primary" message fields to be there and will not be checking the protoc-generated has_xxx flags.   So I can make protoc-c generate something consistent and non-colliding for the optional field tag.  It's simple enough to do.  But it's intrusive, unsightly, and would have to be in more than one place in protoc, but it avoids having to change protobuf3 itself.    

This is certain to not be a generally acceptable solution.   An alternative could be to emulate the C++ approach and replace a has_xxx field with macros.  Perhaps HAS(xxx) and SET(xxx, value), so the user code doesn't need to know the specifics of how the optional field's presence is communicated.  The implementation then can use a variety of means to store the field's presence, including generating unpredictable names that need to be known only to the protobuf-c library and headers.  But it will still break compatibility with the older protobuf-c code.

**Comment by oleg-kiselev** [comment]
I have made changes to my protobuf-c clone that work for me.  But they are tailored to my needs and are not suitable for general use.  But I expect this issue to come up again for others.  The "has_" prefix isn't a problem for protobuf code generated for other languages, and .proto file owners are unlikely to care about protobuf-c users' needs.  This should be addressed by the protobuf-c maintainers.

**Comment by pbor** [comment]
Did you use the current "next" branch? I am going from memory, but I think it does not add has_ fields for proto3 files

**Comment by oleg-kiselev** [comment]
I was using the "master" branch.  I just checked: "next" does work correctly with has_ fields.  

When is the "next" branch going to get merged into "master"?
